dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs

```diff
@@ -557,6 +631,27 @@ await foreach (StreamingChatCompletionsUpdate update in response.ConfigureAwait(
                     continue;
                 }
 
+                try
+                {
+                    // Invoke the pre-invocation filter.
+                    var invokingContext = chatExecutionSettings.ToolCallBehavior?.OnToolInvokingFilter(openAIFunctionToolCall, chat, iteration);
+                    if (invokingContext is not null)
+                    {
+                        // Need to update the chat options in case chat history has changed
+                        this.UpdateChatHistory(chat, chatOptions, chatExecutionSettings);
+
+                        // Check if filter has requested a stop
+                        this.HandleStopBehavior(invokingContext, chatOptions, ref autoInvoke);
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    // Add cancellation message to chat history, turn off tools, and bail out of any remaining tool calls
+                    AddResponseMessage(chatOptions, chat, streamedRole, toolCall, metadata, null, "A tool filter requested cancellation before tool invocation.", this.Logger);
+                    chatOptions.ToolChoice = ChatCompletionsToolChoice.None;
+                    break;
+                }
```

Comment:
This block looks to be repeated , can it be extracted into a method? Same comment for invoked block below.